### PR TITLE
Add validator for transforms

### LINKED ISSUE #246: SYSLGEN: Add warning or error message if view transform does conform to the return type.
SYSLGEN has two parts
- Grammar of the language for which we are generating code for
- Executes a View transform that generates instances of above mentioned grammar types.

### Expected behaviour
Warn user if the view transform does not conform to its return type.

### Actual behaviour
syslgen fails silently and in most cases produces empty text file.

Example Grammar Rule:

```
TopLevelDecl: Comment '\n' (Declaration | FunctionDecl | MethodDecl);
Declaration: VarDecl | ConstDecl | StructType | InterfaceType | AliasDecl;
VarDecl: 'var' identifier '=' TypeName '\n';

```
Checkout the rule `VarDecl` and the transform below
```
  !view varDecl(name <: string, typeName <: string) -> TopLevelDecl:
    name -> (:
      Comment = "// Reference imports to suppress unused errors"    
      Declaration = name -> <Declaration> (:
        VarDecl = name -> <VarDecl> (:
          identifier = name
          TypeName = typeName
        )
      )
    )
```
syslgen should return error in the following cases
- `Comment`, `identifier` and `TypeName` are all required, so error out if not defined.

### LINKED ISSUE #274: Move ValueUtil into a package
`syslutil` seems apropos.

### PR DESCRIPTION
**Pull request by kasunfdo**
- Validate entry point
- Validate filename view
- Validate view content according to return type
- Validate duplicates
- Add unit tests

Fixes #246

Changes proposed in this pull request:
- Add validator to validate transforms

@anz-bank/sysl-developers

### CODE REVIEW

File: sysl2/sysl/exprOp.go [hunk 1]
```diff
@@ -301,69 +303,72 @@ func unionMapSets(lhs, rhs map[string]*sysl.Value) map[string]*sysl.Value {
...
+			syslutil.AppendItemToValueList(setResult.GetSet(), l)
 		}
 	}
 	return setResult
 }
 
-func whereList(txApp *sysl.Application, assign Scope, list *sysl.Value, scopeVar string, rhs *sysl.Expr) *sysl.Value {
-	listResult := MakeValueList()
+func whereList(
+	txApp *sysl.Application, assign syslutil.Scope, list *sysl.Value, scopeVar string, rhs *sysl.Expr) *sysl.Value {
```
Comment (anzdaddy) [inline sysl2/sysl/exprOp.go]: Move `) *sysl.Value {` to a separate line.

File: sysl2/sysl/msg/Message_test.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+		})
+	}
+}
+
+func TestFormatTitle(t *testing.T) {
+	cases := map[string]struct {
+		input    string
+		expected int
+	}{
+		"Length 96": {
```
Comment (sahejsingh) [inline sysl2/sysl/msg/Message_test.go]: test the actual output is correct. not just the length.

File: sysl2/sysl/msg/Message.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+func NewMsg(messageID int, messageData []string) *Msg {
+	return &Msg{MessageID: messageID, MessageData: messageData}
+}
+
+func (m Msg) String() string {
+	message, exists := Messages[m.MessageID]
+
+	if !exists {
+		fmt.Println("[validationError.String] invalid message ID")
```
Comment (anzdaddy) [inline sysl2/sysl/msg/Message.go]: Don't print to stdout. Use logrus at least, but better to panic or return this message as the string value.

File: sysl2/sysl/msg/Message.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+	for i, arg := range m.MessageData {
+		args[i] = arg
+	}
+	return fmt.Sprintf(message, args...)
+}
+
+func (m Msg) genLogMessage() LogMessage {
+	messageType, exists := MessageType[m.MessageID]
+
+	if exists {
```
Comment (anzdaddy) [inline sysl2/sysl/msg/Message.go]:
```go
if messageType, exists := MessageType[m.MessageID]; exists {
```

File: sysl2/sysl/validate/validate.go [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+		if stmt.GetAssign() != nil {
+			if assignCount == 0 && stmt.GetAssign().GetName() != viewName {
+				v.messages[viewName] = append(v.messages[viewName],
+					*msg.NewMsg(msg.ErrMissingReqField, []string{viewName, viewName, "string"}))
+			} else if assignCount > 0 {
+				v.messages[viewName] = append(v.messages[viewName],
+					*msg.NewMsg(msg.ErrExcessAttr, []string{stmt.GetAssign().GetName(), viewName, "string"}))
+			}
+		}
+		assignCount++
```
Comment (anzdaddy) [inline sysl2/sysl/validate/validate.go]: Use `for i, stmt := …` instead of `assignCount`.

File: sysl2/sysl/validate/validate.go [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+			}
+
+		}
+	}
+}
+
+func (v *Validator) compareTuple(
+	specTuple, implTuple *sysl.Type_Tuple,
+	implAttrNames map[string]struct{},
+	viewName, specTupleName string) {
```
Comment (anzdaddy) [inline sysl2/sysl/validate/validate.go]:
```go
	viewName, specTupleName string,
) {
```

File: sysl2/sysl/validate/validate.go [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+			*msg.NewMsg(msg.ErrExcessAttr, []string{attrName, viewName, specTupleName}))
+		delete(implAttrNames, attrName)
+	}
+}
+
+func (v *Validator) compareOneOf(
+	specOneOf *sysl.Type_OneOf,
+	implTuple *sysl.Type_Tuple,
+	implAttrNames map[string]struct{},
+	viewName, specTupleName string) {
```
Comment (anzdaddy) [inline sysl2/sysl/validate/validate.go]:
```go
	viewName, specTupleName string,
) {
```

File: sysl2/sysl/validate/validate.go [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+
```
Comment (anzdaddy) [inline sysl2/sysl/validate/validate.go]: Unnecessary blank line.

File: sysl2/sysl/validate/validate.go [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+			break
+		} else {
+			if _, exists := implAttrs[name]; !exists {
+				matching = false
+			} else {
+				matching = true
+				delete(implAttrNames, name)
+				break
+			}
+		}
```
Comment (anzdaddy) [inline sysl2/sysl/validate/validate.go]:
```go
else if _, matching = implAttrs[name]; !matching {
	delete(implAttrNames, name)
	break
}
```

File: sysl2/sysl/validate/validate.go [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+
+func loadTransform(rootTransform, transformFile string, p *parse.Parser) (*sysl.Application, error) {
+	transform, name := parse.LoadAndGetDefaultApp(rootTransform, transformFile, p)
+
+	if transform == nil {
+		err := errors.New("Unable to load transform")
+		return nil, err
+	}
+
+	return transform.GetApps()[name], nil
```
Comment (anzdaddy) [inline sysl2/sysl/validate/validate.go]:
```go
	if transform, name := parse.LoadAndGetDefaultApp(rootTransform, transformFile, p); transform != nil {
		return transform.GetApps()[name], nil
	}
	return nil, errors.New("Unable to load transform")
```

File: sysl2/sysl/validate/validate.go [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+	tokens[len(tokens)-1] = "sysl"
+	grammarSyslFile := strings.Join(tokens, ".")
+	p := parse.NewParser()
+
+	grammar, name := parse.LoadAndGetDefaultApp(rootGrammar, grammarSyslFile, p)
+	if grammar == nil {
+		err := errors.New("Unable to load grammar-sysl")
+		return nil, err
+	}
+	return grammar.GetApps()[name], nil
```
Comment (anzdaddy) [inline sysl2/sysl/validate/validate.go]: As above

File: sysl2/sysl/validate/validate.go [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+	returnValues.rootTransform = validate.Flag("root-transform",
+		"sysl root directory for input transform file (default: .)").Default(".").String()
+	returnValues.transform = validate.Flag("transform", "grammar.g").Default(".").String()
+	returnValues.grammar = validate.Flag("grammar", "grammar.sysl").Default(".").String()
+	returnValues.start = validate.Flag("start", "start rule for the grammar").Default(".").String()
+	returnValues.loglevel = validate.Flag("log", "log level[debug,info,warn,off]").Default("info").String()
+	returnValues.isVerbose = validate.Flag("verbose", "show output").Short('v').Default("false").Bool()
+
+	return returnValues
+}
```
Comment (anzdaddy) [inline sysl2/sysl/validate/validate.go]:
```go
return &CmdContextParamValidate{
    rootTransform: validate.Flag("root-transform",
        "sysl root directory for input transform file (default: .)").Default(".").String()
    transform:     validate.Flag("transform", "grammar.g").Default(".").String()
    grammar:       validate.Flag("grammar", "grammar.sysl").Default(".").String()
    start:         validate.Flag("start", "start rule for the grammar").Default(".").String()
    loglevel:      validate.Flag("log", "log level[debug,info,warn,off]").Default("info").String()
    isVerbose:     validate.Flag("verbose", "show output").Short('v').Default("false").Bool()
}
```

File: sysl2/sysl/validate/validate.go [hunk 13]
```diff
@@ -113,18 +112,16 @@ func (v *Validator) validateFileName() {
...
-	assignCount := 0
-	for _, stmt := range view.GetExpr().GetTransform().GetStmt() {
+	for i, stmt := range view.GetExpr().GetTransform().GetStmt() {
 		if stmt.GetAssign() != nil {
-			if assignCount == 0 && stmt.GetAssign().GetName() != viewName {
+			if i == 0 && stmt.GetAssign().GetName() != viewName {
 				v.messages[viewName] = append(v.messages[viewName],
 					*msg.NewMsg(msg.ErrMissingReqField, []string{viewName, viewName, "string"}))
-			} else if assignCount > 0 {
+			} else if i > 0 {
```
Comment (anzdaddy) [inline sysl2/sysl/validate/validate.go]: Switch the two tests around and remove `i == 0 && ` from the other one.

File: sysl2/sysl/validate/validate.go [hunk 14]
```diff
@@ -138,7 +135,7 @@ func (v *Validator) validateViews() {
 				v.compareTuple(t.Tuple, resolvedType.GetTuple(),
 					getAttrNames(resolvedType.GetTuple().GetAttrDefs()), viewName, typeName)
 			default:
-				fmt.Println("[validate.validateViews] Unhandled grammar type")
+				logrus.Println("[validate.validateViews] Unhandled grammar type")
```
Comment (anzdaddy) [inline sysl2/sysl/validate/validate.go]: This should be a more specific error classification, like `Error` or `Warn`.

File: sysl2/sysl/validate/validate.go [hunk 15]
```diff
@@ -153,7 +150,7 @@ func (v *Validator) validateViews() {
...
```
Comment (anzdaddy) [inline sysl2/sysl/validate/validate.go]: This should be a more specific error classification, like `Error` or `Warn`.

File: sysl2/sysl/validate/validate.go (hunk 13 above)
Comment (kasunfdo) [reply sysl2/sysl/validate/validate.go]: Fixed

File: sysl2/sysl/validate/validate.go (hunk 14 above)
Comment (kasunfdo) [reply sysl2/sysl/validate/validate.go]: Changed to warn

File: sysl2/sysl/validate/validate.go (hunk 15 above)
Comment (kasunfdo) [reply sysl2/sysl/validate/validate.go]: Changed to warn

File: sysl2/sysl/msg/Message_test.go [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+		t.Run(name, func(t *testing.T) {
+			assert.Equal(t, expected, input.genLogMessage(), "Unexpected result")
+		})
+	}
+}
+
+func TestValidationMsgLogMsg(t *testing.T) {
+	cases := map[string]struct {
+		input Msg
+	}{
```
Comment (anzdaddy) [inline sysl2/sysl/msg/Message_test.go]: nit: Use `map[string}int` and just store message ids.

File: sysl2/sysl/msg/Message_test.go [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+		"Log warning message":   {input: Msg{MessageID: WarnValidatedWithWarn}},
+		"Log info message":      {input: Msg{MessageID: InfoValidatedSuccessfully}},
+		"Log title":             {input: Msg{MessageID: TitleViewName}},
+		"Log unhandled message": {input: Msg{MessageID: 900}},
+	}
+
+	for name, test := range cases {
+		input := test.input
+		t.Run(name, func(t *testing.T) {
+			input.LogMsg()
```
Comment (anzdaddy) [inline sysl2/sysl/msg/Message_test.go]: nit: What is this testing?

File: sysl2/sysl/parse/Parser_test.go [hunk 18]
```diff
@@ -54,11 +56,20 @@ func parseComparable(
 	filename, root string,
 	stripSourceContext bool,
 ) (*sysl.Module, error) {
-	module, err := FSParse(filename, http.Dir(root))
+	module, err := NewParser().FSParse(filename, http.Dir(root))
```
Comment (anzdaddy) [inline sysl2/sysl/parse/Parser_test.go]: nit:
```go
Parse(filename, root)
```

File: sysl2/sysl/parse/parse.go [hunk 19]
```diff
@@ -435,8 +542,65 @@ func postProcess(mod *sysl.Module) {
...
+
+func (p *Parser) GetMessages() map[string][]msg.Msg {
+	return p.Messages
+}
+
+func NewParser() *Parser {
+	return &Parser{
+		AssignTypes: map[string]TypeData{},
+		LetTypes:    map[string]TypeData{},
+		Messages:    map[string][]msg.Msg{}}
```
Comment (anzdaddy) [inline sysl2/sysl/parse/parse.go]: nit: Put the final `}` on its own line.

File: sysl2/sysl/syslutil/typeutil.go [hunk 20]
```diff
@@ -0,0 +1,102 @@
+package syslutil
+
+import sysl "github.com/anz-bank/sysl/src/proto"
+
+// GetTypeDetail returns name of the type and details in string format
+func GetTypeDetail(t *sysl.Type) (string, string) {
```
Comment (anzdaddy) [inline sysl2/sysl/syslutil/typeutil.go]: nit: Make the return values clearer, either through named return values or a more explicit function name like `GetTypeNameAndDetails`.

File: sysl2/sysl/msg/Message_test.go (hunk 17 above)
Comment (kasunfdo) [reply sysl2/sysl/msg/Message_test.go]: Modified test by redirecting log output and comparing against expected values

**Pull request merged**